server/src/server.ts

```diff
@@ -425,6 +407,51 @@ connection.onCompletion((pos: CompletionParams): CompletionItem[] => {
     return completionProvider.getAvailableCompletionItems(pos.textDocument.uri, doc.offsetAt(pos.position));
 });
 
+connection.onDocumentFormatting((_params) => {
+    const doc = documents.get(_params.textDocument.uri);
+    return buildDocumentEditList(doc, yseopmlSettings.documentFormat);
+});
+
+/**
+ * Build and send the list of Text Edit to apply to `document`
+ * accordingly to the settings defined in `documentFormatSettings`.
+ *
+ * @param document the text document to format
+ * @param documentFormatSettings the document format settings to apply
+ */
+export function buildDocumentEditList(document: TextDocument, documentFormatSettings?: IDocumentFormatSettings) {
+    if (documentFormatSettings?.enableDocumentFormat === 'no') {
+        return [];
+    }
+    const inputStream = CharStreams.fromString(document.getText(), document.uri);
+    const lexer = new YmlLexer(inputStream);
+    const tokenStream = new CommonTokenStream(lexer);
+    const parser = new YmlParser(tokenStream);
+    // No need to send syntax errors to the client.
+    parser.removeErrorListeners();
+
+    // Parse the input.
+    const result = parser.kaoFile();
+    const edits: TextEdit[] = [];
+    const visitor = new YmlKaoFileVisitor(
+        new YmlCompletionItemsProvider(),
+        document.uri,
+        new YmlDefinitionProvider(),
+        documentFormatSettings,
+        edits,
+        document,
+    );
+
+    // Visit the result of the parsing.
+    // This fill the edits array.
+    visitor.visit(result);
+    if (parser.numberOfSyntaxErrors > 0) {
```

Comment:
# Question

Can we do this check earlier, or is it `.visit(…)` that allows us to have this number of errors?

Reply:
I'm not sure. I'll do the check and change the code accordingly.

Reply:
Kay.

Reply:
It was OK to avoid the `.visit(…)` call.